packages/grafana-ui/src/components/GraphNG/GraphNG.tsx

```diff
@@ -46,110 +47,101 @@ export const GraphNG: React.FC<GraphNGProps> = ({
-      uniqueScales[scale] = true;
-      scales.push(<Scale key={`scale-${scale}`} scaleKey={scale} />);
-      axes.push(
-        <Axis
-          key={`axis-${scale}-${i}`}
-          scaleKey={scale}
-          label={config.custom?.axis?.label}
-          size={config.custom?.axis?.width}
-          side={config.custom?.axis?.side || AxisSide.Left}
-          grid={config.custom?.axis?.grid}
-          formatValue={v => formattedValueToString(fmt(v))}
-        />
-      );
-    }
-
-    // need to update field state here because we use a transform to merge framesP
-    field.state = { ...field.state, seriesIndex: seriesIdx };
-
-    const colorMode = getFieldColorModeForField(field);
-    const seriesColor = colorMode.getCalculator(field, theme)(0, 0);
-
-    if (customConfig?.line?.show) {
-      seriesGeometry.push(
-        <Line
-          key={`line-${scale}-${i}`}
-          scaleKey={scale}
-          stroke={seriesColor}
-          width={customConfig?.line.show ? customConfig?.line.width || 1 : 0}
-        />
-      );
-    }
+    let { timeIndex } = getTimeField(alignedData);
 
-    if (customConfig?.points?.show) {
-      seriesGeometry.push(
-        <Point key={`point-${scale}-${i}`} scaleKey={scale} size={customConfig?.points?.radius} stroke={seriesColor} />
-      );
-    }
-
-    if (customConfig?.fill?.alpha) {
-      seriesGeometry.push(
-        <Area key={`area-${scale}-${i}`} scaleKey={scale} fill={customConfig?.fill.alpha} color={seriesColor} />
-      );
-    }
-
-    if (seriesGeometry.length > 1) {
-      geometries.push(
-        <SeriesGeometry key={`seriesGeometry-${scale}-${i}`} scaleKey={scale}>
-          {seriesGeometry}
-        </SeriesGeometry>
-      );
+    if (timeIndex === undefined) {
+      timeIndex = 0; // assuming first field represents x-domain
+      builder.addScale({
+        scaleKey: 'x',
+      });
     } else {
-      geometries.push(seriesGeometry);
+      builder.addScale({
+        scaleKey: 'x',
+        isTime: true,
+      });
     }
-
-    if (legend?.isVisible) {
-      legendItems.push({
-        color: seriesColor,
-        label: getFieldDisplayName(field, alignedData),
-        isVisible: true,
-        yAxis: customConfig?.axis?.side === 1 ? 3 : 1,
+    builder.addAxis({
+      scaleKey: 'x',
+      isTime: true,
+      side: AxisSide.Bottom,
+      timeZone,
+      theme,
+    });
+
+    let seriesIdx = 0;
+    const legendItems: LegendItem[] = [];
+    const uniqueScales: Record<string, boolean> = {};
+
+    for (let i = 0; i < alignedData.fields.length; i++) {
+      const field = alignedData.fields[i];
+      const config = field.config as FieldConfig<GraphCustomFieldConfig>;
+      const customConfig = config.custom;
+
+      if (i === timeIndex || field.type !== FieldType.number) {
+        continue;
+      }
+
+      const fmt = field.display ?? defaultFormatter;
+      const scale = config.unit || '__fixed';
+
+      if (!uniqueScales[scale]) {
```

Comment:
can we move this unique scales dictionary to the builder and just have !builder.hasScale(scale)

Reply:
absolutely